### PR TITLE
Add support for Actions Semi ATB1103 MCU and EVB

### PR DESCRIPTION
**Pull request by Mani-Sadhasivam**
Hello,

This PR adds support for Actions Semi ATB1103 MCU and its evaluation board. ATB1103 is part of the  Low power Bluetooth MCUs family (ATB110X) from Actions Semi. The target applications for this MCU are BLE Voice Remote, Smart wearable devices, Smart Home sensors, and BT Mesh device applications. The EVB (Golden board) is designed and manufactured by Actions Semi for their internal use but there are few commercial target boards available for this MCU as well.

This PR is based on the initial code support for the MCU and board developed by Actions Semi and hence the copyrights are shared based on the contributions.

Thanks,
Mani

### CODE REVIEW

**Review by dbkinder — APPROVED** [review]
docs LGTM, thanks.

**Comment by Mani-Sadhasivam** [comment]
@galak Any update on this PR?

**Comment by jhedberg** [comment]
@Mani-Sadhasivam there seems to be a conflict with upstream (with `include/arch/arm/cortex_m/scripts/linker.ld` so could you please rebase? You might also want to take look at whether fixing the warnings from checkpatch makes sense (there are no errors though).

There are also essentially no reviews from codeowners, so could the assigned reviewers please take a look?

**Comment by galak** [comment]
@Mani-Sadhasivam can you address the 'nit' CI check issue.

**Comment by Mani-Sadhasivam** [comment]
@galak Done

File: CODEOWNERS [hunk 1]
```diff
@@ -167,6 +170,8 @@
 /drivers/sensor/lps*/                     @avisconti
 /drivers/sensor/lsm*/                     @avisconti
 /drivers/sensor/st*/                      @avisconti
+/drivers/serial/Kconfig.acts              @Mani-Sadhasivam
+/drivers/serial/uart_acts.c               @Mani-Sadhasivam
```
Comment (galak) [inline CODEOWNERS]: Can you make this just `serial/*acts*`

File: boards/arm/atb1103_golden/Kconfig.board [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2019 Actions (Zhuhai) Technology Co., Ltd
```
Comment (galak) [inline boards/arm/atb1103_golden/Kconfig.board]: Why copyright to Actions?

File: boards/arm/atb1103_golden/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (galak) [inline boards/arm/atb1103_golden/CMakeLists.txt]: Missing Copyright?

**Review by galak — COMMENTED** [review]
Can you shrink the boards/arm/atb1103_golden/doc/img/atb1103_golden.jpg, 4M is kinda big.

**Review by galak — COMMENTED** [review]
The compats for DTS, do you think the names cover other actions semi MCUs?  Just wondering if there is some family name or something to use?

File: drivers/serial/uart_acts.c [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+
+	dev_data->clock_freq = clock_freq;
+
+	/* Set baud rate */
+	err = baudrate_set(dev, dev_data->baud_rate);
+	if (err)
+		return err;
+
+	/* Enable receiver and transmitter */
+	uart->ctrl = ((0x1 << 31) | (0x1 << 30) | (0x1 << 23) | (0x1 << 22) |
```
Comment (galak) [inline drivers/serial/uart_acts.c]: Can this magic number be replaced with defines?

File: drivers/serial/uart_acts.c [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+
+static const struct uart_device_config uart_acts_dev_cfg_0 = {
+	.base = (u8_t *)DT_UART_ACTS_UART_0_BASE_ADDRESS,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.irq_config_func = uart_acts_irq_config_0,
+#endif
+};
+
+static struct uart_acts_dev_data uart_acts_dev_data_0 = {
+	.baud_rate = 115200,
```
Comment (galak) [inline drivers/serial/uart_acts.c]: any reason baud_rate isn't coming from DT?

File: drivers/serial/uart_acts.c [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+
+DEVICE_AND_API_INIT(uart_acts_0, "UART_0", &uart_acts_init,
+		    &uart_acts_dev_data_0, &uart_acts_dev_cfg_0,
+		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		    &uart_acts_driver_api);
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static void uart_acts_irq_config_0(struct device *port)
+{
+	IRQ_CONNECT(IRQ_ID_UART0,
```
Comment (galak) [inline drivers/serial/uart_acts.c]: Any reason IRQ isn't coming from DT?

File: soc/arm/actions/atb110x/Kconfig.defconfig.series [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+if SOC_SERIES_ATB110X
+
+config SOC_SERIES
+	default "atb110x"
+
+config NUM_IRQS
+	# must be >= the highest interrupt number used
+	default 32
+
+config ISR_STACK_SIZE
```
Comment (galak) [inline soc/arm/actions/atb110x/Kconfig.defconfig.series]: Why does this need setting here?  probably a comment at min.

File: soc/arm/actions/atb110x/boot_header.ld [hunk 8]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2019 Manivannan Sadhasivam <mani@kernel.org>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+. = 0xc0;
+LONG(0x00425441);
```
Comment (galak) [inline soc/arm/actions/atb110x/boot_header.ld]: Does this magic number having meaning?  If so a comment about it.

File: soc/arm/actions/atb110x/soc.c [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+	/* Init VD12 */
+	acts_request_vd12_pd(false);
+	acts_request_vd12_largebias(false);
+
+	/* Disable RC_3M */
+	acts_request_rc_3M(false);
+
+	/* Remap Vector base address */
+	sys_write32(0x01000000, VECTOR_BASE);
```
Comment (galak) [inline soc/arm/actions/atb110x/soc.c]: What is magic number 0x01000000?

File: soc/arm/actions/atb110x/soc_irq.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+/**
+ * @file IRQ configuration macros for ATB110X
+ */
+
+#ifndef _ACTIONS_SOC_IRQ_H_
+#define _ACTIONS_SOC_IRQ_H_
+
+#define IRQ_ID_TIMER0         0
```
Comment (galak) [inline soc/arm/actions/atb110x/soc_irq.h]: I'd prefer we remove this file and get values from DT.

File: soc/arm/actions/atb110x/soc_regs.h [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+#define CMU_ANALOG_REG_BASE   0x40002100
+
+#define RTC_REG_BASE          0x40004000
+#define TIMER_REG_BASE        0x40004100
+#define HCL_REG_BASE          0x40004200
+
+#define PMU_REG_BASE          0x40008000
+#define ADC_REG_BASE          0x40008010
+
```
Comment (galak) [inline soc/arm/actions/atb110x/soc_regs.h]: Remove various REG_BASEs that should come from DT.

File: soc/arm/actions/atb110x/soc_reset.h [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+/**
+ * @file Peripheral reset configuration macros for ATB110X
+ */
+
+#ifndef _ACTIONS_SOC_RESET_H_
+#define _ACTIONS_SOC_RESET_H_
+
+#define RESET_ID_DMA		0
```
Comment (galak) [inline soc/arm/actions/atb110x/soc_reset.h]: Should probably encode these in DTS.

**Review by galak — CHANGES_REQUESTED** [review]
Mostly looks good, a number of cleanup comments.

File: boards/arm/atb1103_golden/Kconfig.board (hunk 2 above)
Comment (galak) [reply boards/arm/atb1103_golden/Kconfig.board]: Ah, see you commented in the PR on this.  Is there a reference to their original code?

Comment (Mani-Sadhasivam) [reply boards/arm/atb1103_golden/Kconfig.board]: I have inherited code from them

Comment (Mani-Sadhasivam) [reply boards/arm/atb1103_golden/Kconfig.board]: > Ah, see you commented in the PR on this. Is there a reference to their original code?

No. They just shared a zip file with me based on LTS version and I had to clean a lot for upstream.

**Comment by Mani-Sadhasivam** [comment]
> Mostly looks good, a number of cleanup comments.

Thanks. I'll address them soon.

File: drivers/serial/uart_acts.c (hunk 4 above)
Comment (Mani-Sadhasivam) [reply drivers/serial/uart_acts.c]: Done

File: drivers/serial/uart_acts.c (hunk 5 above)
Comment (Mani-Sadhasivam) [reply drivers/serial/uart_acts.c]: Derived from DT

File: drivers/serial/uart_acts.c (hunk 6 above)
Comment (Mani-Sadhasivam) [reply drivers/serial/uart_acts.c]: Derived from DT

File: soc/arm/actions/atb110x/Kconfig.defconfig.series (hunk 7 above)
Comment (Mani-Sadhasivam) [reply soc/arm/actions/atb110x/Kconfig.defconfig.series]: The stack size is common for all of this SoC series

File: soc/arm/actions/atb110x/boot_header.ld (hunk 8 above)
Comment (Mani-Sadhasivam) [reply soc/arm/actions/atb110x/boot_header.ld]: Added

File: soc/arm/actions/atb110x/soc.c (hunk 9 above)
Comment (Mani-Sadhasivam) [reply soc/arm/actions/atb110x/soc.c]: Comment added

File: soc/arm/actions/atb110x/soc_irq.h (hunk 10 above)
Comment (Mani-Sadhasivam) [reply soc/arm/actions/atb110x/soc_irq.h]: Done

File: soc/arm/actions/atb110x/soc_regs.h (hunk 11 above)
Comment (Mani-Sadhasivam) [reply soc/arm/actions/atb110x/soc_regs.h]: Not all can be removed. But removed the unused ones.

File: soc/arm/actions/atb110x/soc_reset.h (hunk 12 above)
Comment (Mani-Sadhasivam) [reply soc/arm/actions/atb110x/soc_reset.h]: I'm not sure how to pass it since there is no proper reset driver. I'd keep this until we fix later.

**Comment by Mani-Sadhasivam** [comment]
> Can you shrink the boards/arm/atb1103_golden/doc/img/atb1103_golden.jpg, 4M is kinda big.

Done

**Comment by Mani-Sadhasivam** [comment]
> The compats for DTS, do you think the names cover other actions semi MCUs? Just wondering if there is some family name or something to use?

atb110x is the family name, hence added that to board compatible.

**Comment by Mani-Sadhasivam** [comment]
@galak Addressed the comments

**Comment by Mani-Sadhasivam** [comment]
@galak ping! This seems to be stuck for months now

**Review by galak — DISMISSED** [review]
Need to update to new dts macros

**Comment by Mani-Sadhasivam** [comment]
> Need to update to new dts macros

@galak There is no DT gpio used in this PR hence not sure what I should convert. Can you please point to it?

**Comment by Mani-Sadhasivam** [comment]
> Need to update to new dts macros

@galak There is no DT gpio used in this PR hence not sure what I should convert. Can you please point to it?

**Comment by galak** [comment]
> @galak There is no DT gpio used in this PR hence not sure what I should convert. Can you please point to it?

All the DT_INST_{0..9} macros are deprecated and replaced with macros from include/devicetree.h.  Just look most any driver at this point and you'll see something like `DT_INST_PROP(0, foobar)` or the like.

**Comment by Mani-Sadhasivam** [comment]
> DT_INST_

Ah sorry, I misread your comment. Thought you were asking to cleanup DT gpio (it's too late here :P)

**Comment by stephanosio** [comment]
> -:101: WARNING:SPDX_LICENSE_TAG: Missing or malformed SPDX-License-Identifier tag in line 1
> #101: FILE: boards/arm/atb1103_golden/atb1103_golden.dts:1:

Please ignore these warnings (see #24921)

**Comment by Mani-Sadhasivam** [comment]
@carlescufi @galak Ping!

**Comment by Mani-Sadhasivam** [comment]
Hi @carlescufi @galak, Any update on this one as well?

**Comment by carlescufi** [comment]
@ioannisg @galak could you take a look?

**Comment by carlescufi** [comment]
@Mani-Sadhasivam could you please send an email to the devel@ mailing list asking for reviewers for this PR?

**Pull request closed without merging**